src/client/datascience/languageserver/notebookConcatDocument.ts

```diff
@@ -0,0 +1,150 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+import * as path from 'path';
+import * as uuid from 'uuid/v4';
+import {
+    CodeAction,
+    CodeActionContext,
+    CodeLens,
+    Command,
+    CompletionItem,
+    CompletionList,
+    Diagnostic,
+    DiagnosticRelatedInformation,
+    DocumentHighlight,
+    DocumentSelector,
+    DocumentSymbol,
+    EndOfLine,
+    Hover,
+    Location,
+    LocationLink,
+    NotebookConcatTextDocument,
+    NotebookDocument,
+    Position,
+    Range,
+    SymbolInformation,
+    TextDocument,
+    TextDocumentChangeEvent,
+    TextDocumentContentChangeEvent,
+    TextEdit,
+    Uri,
+    WorkspaceEdit
+} from 'vscode';
+import { IVSCodeNotebook } from '../../common/application/types';
+import { HiddenFileFormatString } from '../../constants';
+
+/**
+ * This helper class is used to present a converted document to an LS
+ */
+export class NotebookConcatDocument implements TextDocument {
+    public get notebookUri() {
+        return this.notebookDoc.uri;
+    }
+
+    public get uri() {
+        return this.dummyUri;
+    }
+
+    public get fileName() {
+        return this.dummyFilePath;
+    }
+
+    public get isUntitled() {
+        return this.notebookDoc.isUntitled;
+    }
+
+    public get languageId() {
+        return this.notebookDoc.languages[0];
+    }
+
+    public get version() {
+        return this._version;
+    }
+
+    public get isDirty() {
+        return this.notebookDoc.isDirty;
+    }
+
+    public get isClosed() {
+        return this.concatDocument.isClosed;
+    }
+    public get eol() {
+        return EndOfLine.LF;
+    }
+    public get lineCount() {
+        return this.notebookDoc.cells.map((c) => c.document.lineCount).reduce((p, c) => p + c);
+    }
+    public concatDocument: NotebookConcatTextDocument;
+    private dummyFilePath: string;
+    private dummyUri: Uri;
+    private _version = 1;
+    constructor(private notebookDoc: NotebookDocument, notebookApi: IVSCodeNotebook, selector: DocumentSelector) {
+        const dir = path.dirname(notebookDoc.uri.fsPath);
+        this.dummyFilePath = path.join(dir, HiddenFileFormatString.format(uuid().replace(/-/g, '')));
+        this.dummyUri = Uri.file(this.dummyFilePath);
+        this.concatDocument = notebookApi.createConcatTextDocument(notebookDoc, selector);
+        this.concatDocument.onDidChange(this.onDidChange.bind(this));
```

Comment:
```suggestion
        this.concatDocument.onDidChange(this.onDidChange, this, disposables);
```

TIp: you can pass `this` as the argument as well, instead of using `bind` (not a required change).
However, we need to dispose this event handler.